File: tyama_aizu0115.rb
#!/usr/bin/ruby
require 'rational'
def gauss(a)
	if !a || a.length==0 then return nil end
	n=a.length
	n.times{|i|
		if a[i][i]==0
			if (i+1).step(n-1){|j|
				if a[j][i]!=0
					i.step(n){|k| a[i][k]+=a[j][k]}
					break
				end
			}
				return nil
			end
		end
		n.times{|j|
			if j!=i then
				r = a[j][i].is_a?(Rational) || a[i][i].is_a?(Rational) ? a[j][i]/a[i][i] : Rational(a[j][i],a[i][i])
				i.step(n){|k| a[j][k] = a[j][k] - a[i][k]*r}
			end
		}
	}

	n.times{|i|
		x=Rational(a[i][i],1)
		a[i].length.times{|j|
			a[i][j] /= x
		}
	}

	return a
end

uaz=gets.split.map(&:to_i)
enemy=[0]+gets.split.map(&:to_i).zip(uaz).map{|x,y|-x+y}
b0=[1]+gets.split.map(&:to_i).zip(uaz).map{|x,y|x-y}
b1=[1]+gets.split.map(&:to_i).zip(uaz).map{|x,y|x-y}
b2=[1]+gets.split.map(&:to_i).zip(uaz).map{|x,y|x-y}
sol=gauss([b0,b1,b2,enemy,[1,0,0,0]].transpose)
puts sol && sol.map(&:last).all?{|e|0<=e&&e<=1} ? :MISS : :HIT

File: aizu/tyama_aizu0115.py
#!/usr/bin/python
from fractions import Fraction
import sys
if sys.version_info[0]>=3: raw_input=input

def gauss(a):
	if not a or len(a)==0: return None
	n=len(a)
	for i in range(n):
		if a[i][i]==0:
			for j in range(i+1,n):
				if a[j][i]!=0:
					for k in range(i,n+1): a[i][k]+=a[j][k]
					break
			else:
				return None
		for j in range(n):
			if i!=j:
				r = Fraction(a[j][i],a[i][i])
				for k in range(i,n+1): a[j][k] = a[j][k] - a[i][k]*r
	for i in range(n):
		x=Fraction(a[i][i],1)
		for j in range(len(a[i])):
			a[i][j] /= x
	return a

uaz=list(map(int,raw_input().split()))
enemy=[0]+[-x+y for x,y in zip(map(int,raw_input().split()),uaz)]
b0=[1]+[x-y for x,y in zip(map(int,raw_input().split()),uaz)]
b1=[1]+[x-y for x,y in zip(map(int,raw_input().split()),uaz)]
b2=[1]+[x-y for x,y in zip(map(int,raw_input().split()),uaz)]
sol=gauss(list(map(list,zip(b0,b1,b2,enemy,[1,0,0,0]))))
print('MISS' if sol and all(0<=e[-1]<=1 for e in sol) else 'HIT')